src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -39,13 +39,361 @@ namespace std
+        case 7:
+            color += PC.Ks*pow(VdotR, PC.Ns)*Fresnel_dielectric(PC.Ni, NdotL);
+            break;
+        default:
+            break;
+        }
+    }
+    else color = PC.Kd;
+
+    OutColor = vec4(color, 1.0);
+}
+)";
+    constexpr const char* FRAG_SHADER_NO_UV_PBR =
+R"(#version 430 core
+
+layout (location = 0) in vec3 LocalPos;
+layout (location = 1) in vec3 ViewPos;
+layout (location = 2) in vec3 Normal;
+layout (location = 0) out vec4 OutColor;
+
+layout (push_constant) uniform Block {
+    vec3 ambient;
+    vec3 albedo;//MTL's diffuse
+    vec3 specular;
+    vec3 emissive;
+    vec4 Tf;//w component doesnt matter
+    float shininess;
+    float opacity;
+    float bumpFactor;
+    //PBR
+    float ior;
+    float roughness;
+    float metallic;
+    float sheen;
+    float clearcoatThickness;
+    float clearcoatRoughness;
+    float anisotropy;
+    float anisoRotation;
+    //extra info
+    uint extra;
+} PC;
+
+#define PI 3.14159265359
+#define FLT_MIN 1.175494351e-38
+
+#include <irr/builtin/glsl/brdf/diffuse/oren_nayar.glsl>
+#include <irr/builtin/glsl/brdf/specular/ndf/ggx_trowbridge_reitz.glsl>
+#include <irr/builtin/glsl/brdf/specular/geom/ggx_smith.glsl>
+#include <irr/builtin/glsl/brdf/specular/fresnel/fresnel.glsl>
+
+void main()
+{
+    vec3 N = normalize(Normal);
+    //some approximation for computing tangents without UV
+    vec3 c1 = cross(N, vec3(0.0, 0.0, 1.0));
+    vec3 c2 = cross(N, vec3(0.0, 1.0, 0.0));
+    vec3 T = (dot(c1,c1) > dot(c2,c2)) ? c1 : c2;
+    T = normalize(T);
+    vec3 B = normalize(cross(N,T));
+    vec3 V = -ViewPos;
+
+    vec3 NdotV = dot(N,V);
+#define NdotL NdotV
+#define NdotH NdotV
+
+    vec3 color = PC.emissive*0.01;
+    if (NdotL > FLT_MIN)
+    {
+        float lightDistance2 = dot(V,V);
+        float Vrcplen = inversesqrt(lightDistance2);
+        NdotV *= Vrcplen;
+        V *= Vrcplen;
+
+        vec3 TdotV = dot(T,V);
+        vec3 BdotV = dot(B,V);
+#define TdotL TdotV
+#define BdotL BdotV
+#define TdotH TdotV
+#define BdotH BdotV
+
+        float at = sqrt(PC.roughness);
+        float ab = at*(1.0 - PC.anisotropy);
+
+        float fr = Fresnel_dielectric(PC.ior, NdotV);
+        float one_minus_fr = 1.0-fr;
+        float diffuseFactor = 1.0 - one_minus_fr*one_minus_fr;
+        float diffuse = 0.0;
+        if (PC.metallic < 1.0)
+        {
+            if (PC.roughness==0.0)
+                diffuse = 1.0/PI;
+            else
+                diffuse = oren_nayar(PC.roughness, N, V, V, NdotL, NdotV);
+        }
+        float specular = 0.0;
+        if (NdotV > FLT_MIN)
+        {
+            float ndf = GGXBurleyAnisotropic(PC.anisotropy, PC.roughness, TdotH, BdotH, NdotH);
+            float geom = GGXSmithHeightCorrelated_aniso_wo_numerator(at, ab, TdotL, TdotV, BdotL, BdotV, NdotL, NdotV);
+            specular = ndf*geom*fr;
+        }
+
+        color += (diffuseFactor*diffuse*PC.albedo + specular) * NdotL / lightDistance2;
+    }
```

Comment:
you should gather BRDF inputs and then call specific whole `brdf_cos_eval` functions for given PBR BRDF

Reply:
i think i could use some simple example of what u have in mind

Reply:
discussed in discord.